capstone_project/src/main/java/com/google/sps/servlets/MapServlet.java

```diff
@@ -28,52 +38,56 @@
 
-  private void addUWToSchools() {
-    ArrayList<String> uwNews =
-        new ArrayList<String>(
-            Arrays.asList(
-                "https://www.washington.edu/coronavirus/2020/07/10/back-to-school-town-hall/",
-                "https://www.washington.edu/news/2020/07/16/wsas-2020/?utm_source=UW%20News&utm_medium=tile&utm_campaign=UW%20NEWS",
-                "https://www.washington.edu/news/2020/07/29/expert-faq-wildfires-in-the-pacific-northwest-during-the-covid-19-pandemic/"));
-    School uw = new School("UW", 47.655277, -122.303606, uwNews);
-    schools.add(uw);
-  }
+    // Load Schools from Datastore.
+    Query query = new Query("School");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
 
-  private void addUWBToSchools() {
-    ArrayList<String> uwbNews =
-        new ArrayList<String>(
-            Arrays.asList(
-                "https://news.uci.edu/2020/06/30/virtual-nurses-make-a-real-difference/",
-                "https://news.uci.edu/2020/07/07/uci-chancellor-emeritus-michael-v-drake-named-university-of-california-president/",
-                "https://news.uci.edu/2020/06/18/uci-podcast-jessica-millward-on-the-meaning-and-importance-of-juneteenth/"));
-    School uwb = new School("UWB", 47.758821, -122.190725, uwbNews);
-    schools.add(uwb);
-  }
+    for (Entity entity : results.asIterable()) {
+      String schoolName = (String) entity.getProperty("name");
+      double schoolLatitude = (double) entity.getProperty("latitude");
+      double schoolLongitude = (double) entity.getProperty("longitude");
+      School retrievedSchool = new School(schoolName, schoolLatitude, schoolLongitude);
+      schoolsSet.add(retrievedSchool);
+    }
 
-  /*Add some hard-coded School objects to the array of schools that will later be converted into markers on the map.*/
-  public MapServlet() {
-    addUCIToSchools();
-    addUWToSchools();
-    addUWBToSchools();
+    // Send the School objects retrieved from Datastore.
+    response.setContentType("application/json;");
+    Gson gson = new Gson();
+    response.getWriter().println(gson.toJson(schoolsSet));
   }
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("application/json;");
-    Gson gson = new Gson();
-    response.getWriter().println(gson.toJson(schools));
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Initialize datastore object.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get the input from the school entry form.
+    String schoolName = request.getParameter("name-input");
+    double schoolLatitude = Double.parseDouble(request.getParameter("latitude-input"));
+    double schoolLongitude = Double.parseDouble(request.getParameter("longitude-input"));
```

Comment:
I kind of liked the old method you had because it had a null check for these. But like Lynda said, you didn't use a default string. I think you could either bring back the old method and remove the default parameter and just return an empty string if the value if null or you could a null check for these getParameter things here in your doPost method.

Reply:
@Leoism fixed!